### PR TITLE
Time is Weird

### PR DESCRIPTION
**Pull request by PandaZ3D**
A draft of a set of abstractions for user space and the kernel.

### CODE REVIEW

File: text/0000-time.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+```rust
+struct Clock {
+    info: ClockInfo,
+    usage: ClockUsage,
+    id: u64 // leads us to the clock source
+}
+```
+`ClockInfo` is meant to describe a clock to users. It contains the last value read from the clock and properties of the clock, such as resolution.
+
+```rust
```
Comment (gvnn3) [inline text/0000-time.md:63]: Please specify the meanings of each of these fields.  I suspect some are time (nano/femto/pico seconds) and one is drift in parts per million.  Check out this page fore more on this: http://www.ntp.org/ntpfaq/NTP-s-sw-clocks-quality.htm

**Review by gvnn3 — CHANGES_REQUESTED** [review]

File: text/0000-time.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+## Clock and ClockInfo
+
+To support the above functionality, Twizzler exposes a `Clock` abstraction to user space, and APIs revolve around this and `ClockInfo`. A `Clock` is a logical abstraction exposed to users that serves a particular purpose and ticks at a particular pace.
+
+```rust
+struct Clock {
+    info: ClockInfo,
+    usage: ClockUsage,
+    id: u64 // leads us to the clock source
```
Comment (dbittman) [inline text/0000-time.md]: I'd like this to be a type, something like
`#[repr(transparent)] struct ClockId(u64)`

File: text/0000-time.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+```
+The `value` is a `Duration` type, and thus, `ClockInfo` can be thought of as a POSIX `timespec`, but much better. It provides descriptions of what is being read, not just the value. This additional metadata is essential for scientific applications that want to use time as a tool for measuring some event. Applications need to know about the properties of clocks to get accurate measurements and reduce errors in experiments.
+
+`ClockUsage` is a set of enums describing the purpose of the clock. This gives users control of what type of clock they are reading from (when talking to the kernel), and they can expect certain invariants to be maintained, such as a clock being monotonic.
+
+```rust
+enum ClockUsage {
+    Monotonic,
+    RealTime,
+    Unknown
```
Comment (dbittman) [inline text/0000-time.md]: Place Unknown as first element.

File: text/0000-time.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+}
+```
+A few things here. For starters, `sys_read_clock_info`, a new system call. We pass in the type of clock we want (`ClockUsage`), an optional `clock_id`, and some flags. By default, this returns a `ClockInfo` object with all fields filled in. The returned `ClockInfo`, in this case, is generated from a `ClockUsage` specific clock. We will get to the use of `clock_id` later.
+
+Semantically, this is the same program, except it uses different abstractions. The `ClockInfo` type has a set of methods to return internal values. We calculate the elapsed time by subtracting two `Duration` types that sampled different points in time from the same clock. Internally, `std::time` does something like this using the `Instant` type.
+
+Twizzler exposes a new set of system calls related to timekeeping. Other than `sys_read_clock_info`, which is helpful in reading a clock and learning about its properties, users need ways to steer a clock and prevent drift. Users also need support to discover available clocks.
+
+```rust
+fn sys_read_clock_info(clock: ClockUsage, clock_id: u64, flags: ReadClockFlags) -> Result<ClockInfo, ReadClockError>
```
Comment (dbittman) [inline text/0000-time.md]: As we discussed on wednesday, I'd like to see this changed to an API that doesn't admit erroneous combinations of arguments.

**Review by dbittman — CHANGES_REQUESTED** [review]
See comments. I'd also suggest removing references to time adjustment and specifying that a future RFC will define the API to adjust clocks.

**Comment by PandaZ3D** [comment]
Thanks for the suggestions everyone. I have incorporated feedback from our conversations. 

The only new thing is the introduction of `TimeSpan` meant to represent some point in time. It is a timestamp like `Duration`/`timespec`, except unlike legacy abstractions, is designed to support high precision time sources. `TimeSpan` uses two 64-bit integers encapsulated behind `Seconds` and `FemtoSeconds`.

**Comment by dbittman** [comment]
Is the rationale that the TimeSpan is more future-proof in that is supports femtosecond resolution, and that the other precision information is given in femtoseconds anyway, so it typechecks better than using nanoseconds?

That's fine by me. We'll, of course, need to impl From<TimeSpan> for Duration (or, if that's not possible, impl Into<Duration> for TimeSpan), since the Rust std library will still use Duration.

**Review by dbittman — APPROVED** [review]
Looks good!

**Comment by PandaZ3D** [comment]
Yes. Femtosecond clock hardware exists, but is pricey. However, it is reasonable to assume that modern hardware could have picosecond resolution or precision. If we feel that femtoseconds is too fine grained, we can go with picoseconds as well. Nanoseconds is too coarse grained which is a nonstarter.

**Pull request merged**